### ISSUE TITLE
Is there any way to retrieve the yhat (in-sample forecast) estimated in the training data set?

### ISSUE THREAD

**Issue by raphaccf** [pycaret/pycaret#1999]
### Discussed in https://github.com/pycaret/pycaret/discussions/1985

<div type='discussions-op-text'>

<sup>Originally posted by **raphaccf** December 22, 2021</sup>
Hi there!!

It is possible to get the yhat (the estimated target variable) from predict_model(). However, this predict function only works for the test period. 

Is there any way to recover yhat from training period? 

The evaluation metrics are already calculated in several functions (compare_models; create_model; tuned_model;  blend_model), but I couldn't find any way, if it is possible, to have the yhat's. Plots of some models ( like arima, for example) also bring yhat, but not for other models, and could not find how to recovery even for such method. 
It would be important for analysis and forecast intervals.

Kind regards!</div>

**Comment by ngupta23** [comment]
You can do something like this. This will return the data for insample predictions. 

```python
data = exp.plot_model(my_model, plot="residuals", return_data=True)
```

Will this work?

**Comment by raphaccf** [comment]
Yes, this works for some models, like auto.arima and theta (but setting "return_fig=False" does not avoid the plot to be shown), However, this does not work for other models (the majority, I guess) , and the following message is presented:
 "In sample predictions has not been implemented for this estimator of type 'BaseCdsDtForecaster' in `sktime`. When this is implemented, it will be enabled by default in pycaret." 
So, we can retrive such information from plot functions, but just for some models (until now), although for every model calculations for evaluation metrics are made (_compete_models()_ or _create_model()_, for example), then we have 'y' (real) and 'yhat' (prediciton in-sample) somewhere in the code process, but we can not retrive 'yhat' (the solution is to use plot_model(), but it is working for some models).

code
`from pycaret.time_series import *
from pycaret.datasets import get_data
data = get_data('airline', verbose=True)
setup(data, fh = 6,fold = 3,fold_strategy='expanding',seasonal_period=12,session_id = 123,verbose=False)
lgbm = create_model('lightgbm_cds_dt',verbose=False)
plot=plot_model(estimator = lgbm, plot = 'residuals',return_data=True,return_fig=False,verbose=False)
plot`

**Comment by ngupta23** [comment]
> ... although for every model calculations for evaluation metrics are made (compete_models() or create_model(), for example), then we have 'y' (real) and 'yhat' (prediciton in-sample) somewhere in the code process

@raphaccf There is a misunderstanding here. The `yhat` values that we get during create_model/tune_model are all out of sample values. We use part of each fold in the training data to "train" and part of it to "cross_validate". This is not the same as "in-sample" y-hat for the model that you get at the end of `create_model` (which is trained on the entire training dataset). If you do `plot_model(plot="cv")`, you can see how the training data is split into folds and what part of the fold is used for training and what part for the cross-validation metrics.

> However, this does not work for other models (the majority, I guess) , and the following message is presented:
"In sample predictions has not been implemented for this estimator of type 'BaseCdsDtForecaster' in sktime. When this is implemented, it will be enabled by default in pycaret."

The is a correct observation. In-sample residuals is not supported by reduced regression based models and a few others. It would be best to file an issue with `sktime` to see if this can be incorporated in the library. Once this is done, pycaret time series module will pick it up automatically.

In the mean time...If you want to get in-sample predictions (**for models that support it**) without using `plot_model`, you can do something like this (using the OOP interface only for now)

```python
model = exp.create_model("ets")
exp.get_residuals(model)
```

Also, if by in-sample residuals, you really meant the out-of-sample residuals from the cross-validation on the training data, then let me know. We can open a new issue for this.

**Comment by raphaccf** [comment]
Thanks for your response. I understand your points and especially the limitation that comes from sktime models. My main goal with residuals (in-sample predicted values ​​and actual values, retrived from plot_model as you metioned ​​) is to build prediction intervals, so I need them (for models that don't create intervals in their estimation process) .

**Comment by ngupta23** [comment]
If you want to use only models that provide a prediction interval, you can refer to this FAQ: https://github.com/pycaret/pycaret/discussions/1814
